src/build/build.ts

```diff
@@ -40,105 +40,100 @@ export interface BuildOptions {
-    let unbundledPostProcessing = Promise.all([loadSWConfig, waitFor(unbundledPhase)]).then((results) => {
-      let swConfig: SWConfig = results[0];
-      return addServiceWorker({
-        buildRoot: 'build/unbundled',
-        project: polymerProject,
-        swConfig: swConfig,
-      });
-    });
+  logger.info(`Building application...`);
+
+  logger.debug(`Reading source files...`);
+  let sourcesStream = polymerProject.sources()
+    .pipe(polymerProject.splitHtml())
+    .pipe(gulpif(/\.js$/, new JSOptimizeStream(optimizeOptions.js)))
+    .pipe(gulpif(/\.css$/, new CSSOptimizeStream(optimizeOptions.css)))
+    .pipe(gulpif(/\.html$/, new HTMLOptimizeStream(optimizeOptions.html)))
+    .pipe(polymerProject.rejoinHtml());
+
+  logger.debug(`Reading dependencies...`);
+  let depsStream = polymerProject.dependencies()
+    .pipe(polymerProject.splitHtml())
+    .pipe(gulpif(/\.js$/, new JSOptimizeStream(optimizeOptions.js)))
+    .pipe(gulpif(/\.css$/, new CSSOptimizeStream(optimizeOptions.css)))
+    .pipe(gulpif(/\.html$/, new HTMLOptimizeStream(optimizeOptions.html)))
+    .pipe(polymerProject.rejoinHtml());
+
+  let buildStream = mergeStream(sourcesStream, depsStream)
+    .once('data', () => { logger.debug('Analyzing build dependencies...'); })
+    .pipe(polymerProject.analyzer);
+
+  let unbundledPhase = forkStream(buildStream)
+    .once('data', () => { logger.info('Generating build/unbundled...'); })
+    .pipe(
+      gulpif(
+        options.insertDependencyLinks,
+        new PrefetchTransform(polymerProject.root, polymerProject.entrypoint,
+          polymerProject.shell, polymerProject.fragments,
+          polymerProject.analyzer)
+      )
+    )
+    .pipe(dest('build/unbundled'));
+
+  let bundledPhase = forkStream(buildStream)
+    .once('data', () => { logger.info('Generating build/bundled...'); })
+    .pipe(polymerProject.bundler)
+    .pipe(dest('build/bundled'));
+
+  let swPrecacheConfig = path.resolve(polymerProject.root, options.swPrecacheConfig || 'sw-precache-config.js');
+  let loadSWConfig = parsePreCacheConfig(swPrecacheConfig);
+
+  loadSWConfig.then((swConfig) => {
+    if (swConfig) {
+      logger.debug(`Service worker config found`, swConfig);
+    } else {
+      logger.debug(`No service worker configuration found at ${swPrecacheConfig}, continuing with defaults`);
+    }
+  });
 
-    // Once the bundled build stream is complete, create a service worker for the build
-    let bundledPostProcessing = Promise.all([loadSWConfig, waitFor(bundledPhase)]).then((results) => {
-      let swConfig: SWConfig = results[0];
-      return addServiceWorker({
-        buildRoot: 'build/bundled',
-        project: polymerProject,
-        swConfig: swConfig,
-        bundled: true,
-      });
+  // Once the unbundled build stream is complete, create a service worker for the build
+  let unbundledPostProcessing = Promise.all([loadSWConfig, waitFor(unbundledPhase)]).then((results) => {
```

Comment:
I don't suppose it would be any nicer to use an IIAFE?

```typescript
(async () => {
  const result = await Promise.all([loadSWConfig, waitFor(unbundledPhase)]);
  const swConfig = result[0];
  await addServiceWorker(...);
}());
```

Just brainstorming to see if there's any nice pattern for doing all this parallel work...